docs/2019-03-28-All-Path-Reachability-Proofs.md

```diff
@@ -0,0 +1,273 @@
+which can be written as `t ∧ p` where the interpretation of `t` contains at most
+one element and `p` is a predicate.
+
+_Extended constructor patterns_ will be those extended function-like patterns
+for which `t` is a functional term, composed out of constructor-like symbols
+and variables.
+
+
+Definitions
+-----------
+
+### Weak always
+
+Given a formula `φ`, let `[w]φ` denote the formula “weak always” `φ`,
+defined by:
+
+```
+[w]φ := νX.φ ∨ (○X ∧ •⊤)
+```
+
+one consequence of the above is that `[w]φ = φ ∨ (○[w]φ ∧ •⊤)`.
+
+Given this definition of weak always, an all-path reachability claim
+```
+∀x.φ(x) → [w]∃z.ψ(x,z)
+```
+basically states that if `φ(x)` holds for a configuration `γ`, for some `x`,
+then `P(γ)`, where `P(G)` is recursively defined on configurations as:
+* there exists `z` such that `ψ(x,z)` holds for `G`
+* or
+    * `G` is not stuck (`G → • T`) and
+    * `P(G')` for all configurations `G'` in which `G` can transition
+
+
+
+Problem Description
+-------------------
+
+Given a set of all-path reachability claims, of the form `∀x.φ(x) → [w]∃z.ψ(x,z)`,
+we are trying to prove all of them together.
+
+
+Algorithms
+----------
+(detailed in the next section)
+
+
+### Algorithm `proveAllPath`
+
+__Input:__ claims `∀x₁.φ₁ → [w]∃z₁.ψ₁`, `∀x₂.φ₂ → [w]∃z₂.ψ₂`, …, `∀xₙ.φₙ → [w]∃zₙ.ψₙ`
+
+__Output:__ Proved or Unproved
+
+* For each claim `∀x.φ → [w]∃z.ψ`
+    * Let `Goals := { ∀x.φ → [w]∃z.ψ }`
+    * While `Goals` is not empty:
+        * Extract and remove `goal` of the form `∀x.φ → [w]∃z.ψ` from `Goals`
+        * Let `goalᵣₑₘ := ∀x. (φ ∧ ¬∃z.⌈φ ∧ ψ⌉) → [w]∃z.ψ`
+        * If `goalᵣₑₘ` is trivialy valid (i.e., if `φ ∧ ¬∃z.⌈φ ∧ ψ⌉ ≡ ⊥`)
+            * continue to the next goal
+        * If not the first while iteration for this claim:
+            * `(Goals', goal'ᵣₑₘ) := deriveSeq(goalᵣₑₘ, claims)`
+                (alternatively, one can use `(Goals', goal'ᵣₑₘ) := derivePar(goalᵣₑₘ, claims)`)
+        * Else, let `(Goals', goal'ᵣₑₘ) := (∅, goalᵣₑₘ)`
+        * Let `(Goals'', goal''ᵣₑₘ) := derivePar(goal'ᵣₑₘ, axioms)`
+        * If `goal''ᵣₑₘ` is not trivially valid (its lhs is not equivalent to `⊥`)
+            * Return `Unprovable`
+        * Let `Goals := Goals ∪ Goals' ∪ Goals''`
+* Return `Provable`
+
+__Note:__ Since the derivation process can continue indefinitely, one could add
+a bound on the total number of (levels of) expansions attempted before
+returning `Unprovable`.
+
+### Algorithm `derivePar`
+
+__Input:__: goal and set of claims/axioms
+
+* goal: `∀x.φ → [w]∃z.ψ`
+* Either claims or axioms:
+    * claims `∀x₁.φ₁ → [w]∃z₁.ψ₁`, `∀x₂.φ₂ → [w]∃z₂.ψ₂`, …, `∀xₙ.φₙ → [w]∃zₙ.ψₙ`
+    * axioms `∀x₁.φ₁ → •∃z₁.ψ₁`, `∀x₂.φ₂ → •∃z₂.ψ₂`, …, `∀xₙ.φₙ → •∃zₙ.ψₙ`
+    * we will not consider the form, but just the patterns and variables involved
+
+__Output:__ `(Goals, goalᵣₑₘ)`
+
+* Let `goalᵣₑₘ := ∀x.(φ ∧ ¬∃x₁.⌈φ∧φ₁⌉ ∧ …  ∧ ¬∃xₙ.⌈φ∧φₙ⌉) → [w]∃z.ψ`
```

Comment:
Similarly, we can prove that `φ ∧ ¬∃x₁.⌈φ∧φ₁⌉ ∧ …  ∧ ¬∃xₙ.⌈φ∧φₙ⌉` equals `φ ∧ ¬∃x₁.φ₁ ∧ …  ∧ ¬∃xₙ.φₙ`, given that `φ` is a function-like pattern.